### DISCUSSION TITLE
[RFC] Changing definition of `Comment`

### DISCUSSION THREAD

**Discussion by kdy1** [swc-project/swc#3407] [RFC]
I'm considering changing `Comment` to store `Arc<String>` instead of `String`.
This is because comments are typically simply passed around, and there are lots of cases where comments should be cloned.
Also, it's quite hard to modify the content of `Comment` without allocation, so it's not likely to degrade performance.

**Comment by kdy1** [comment]
cc @dsherret @bartlomieju 
cc @devongovett @mischnic 

Do you use some methods like `to_ascii_lowercase` or something similar method, where the output has same length as input? (If length is different, you are re-allocating anyway)

**Comment by dsherret** [comment]
When do you need to clone them? Is this something that could be done if desired after parsing? I think it’s not a big deal whatever is decided.

**Reply by kdy1** [reply]
I needed to clone them while translating ast into a different format.

**Reply by kdy1** [reply]
Also, some comments like `PURE` are cloned lots of times.

**Reply by dsherret** [reply]
One possibility is after parsing, the comments could be taken via something like `take_all` (https://github.com/swc-project/swc/blob/db674354d3ee658c0b2dfebee814218a2e647cf9/crates/swc_common/src/comments.rs#L490) and then mapped to be in an `Arc` if necessary at that point. Whatever is decided should be ok for us so no big deal whatever way.

By the way, maybe https://github.com/swc-project/swc/issues/1952 should be thought about at the same time. Perhaps passing around `Arc<Comment>` would reduce the cloning necessary in that API (ex. for `get_x` methods https://github.com/swc-project/swc/blob/db674354d3ee658c0b2dfebee814218a2e647cf9/crates/swc_common/src/comments.rs#L41).

Edit: Sorry, I’m on my phone too lazy to fix my messy comment 😅

**Reply by kdy1** [reply]
Oh, yes you are right. I can change type when necessary.

Also yes, I forgot to mention it, but actually the api of `Comments` is also problematic because of cloning.
